### ISSUE TITLE
OKSloss and keypoints detection results

### ISSUE THREAD

**Issue by coordxyz** [TexasInstruments/edgeai-yolov5#10]
## ❔Question
Hi, I'm trying to implement keypoint detection according to your paper and official yolov5-v6.1. But I have some problems and can't find what they are. Hope you can give me some advices. Thx~
Here are some observations.
1. It seems that my 17 keypoint positions deform to the center of person bounding box when using OKSloss.
Results are generated by the model trained with OKSloss. I implement the OKSloss based on COCOAPI.
![image](https://user-images.githubusercontent.com/47654539/168287603-fd301500-0bd5-41ac-bd03-7bf05c3c2ee1.png)
![image](https://user-images.githubusercontent.com/47654539/168288236-16c2caa9-17e7-4485-8a49-a54fb858aa57.png)
![image](https://user-images.githubusercontent.com/47654539/168288547-c08ad14a-8b06-4882-8895-76788f41e564.png)

2. At first I thought it was caused by bugs in building targets. But the results generated by the model trained with L1loss make me more confused. Even though the keypoint positions are not correct, but they are seperated from each other.
![image](https://user-images.githubusercontent.com/47654539/168288699-ac497eb3-13a5-491b-b28d-a562fcb588d0.png)
![image](https://user-images.githubusercontent.com/47654539/168289199-938b0af5-485e-42f5-a2a9-28e5fc2a99c1.png)
![image](https://user-images.githubusercontent.com/47654539/168289262-0f658426-48e2-4cb6-8e33-4ebf607c169e.png)

3. The keypoint position regression loss (both OKS and L1) is constant. But the keypoint visiablity loss decline looks normal.
kpt->OKSloss, kvis->BCE
![image](https://user-images.githubusercontent.com/47654539/168292317-59c14f8b-9215-48ab-95a7-81288fe4a4e1.png)
kpt->L1loss, kvis->BCE
![image](https://user-images.githubusercontent.com/47654539/168292254-79ed7470-d31f-48f8-ab74-1048b538fb0c.png)

4. I simply add a keypoint detection layer after the object detection layer in yolov5m network structure. I initialize the network with the pretrained yolov5m model and without freezing any layers. The object detection results look normal as shown below.
Results with OKSloss-keypoints
![image](https://user-images.githubusercontent.com/47654539/168293775-58bb2ce5-f3f8-4518-a543-7a4a2f07cd82.png)
Resutls with L1loss-keypoints
![image](https://user-images.githubusercontent.com/47654539/168293862-b31d9b6e-c3ce-41b2-93da-add8a6f5b32f.png)
I also tried add an extra head for keypoint detection but the results are similar.

**Comment by coordxyz** [comment]
Fix some bugs and get normal results with L1loss as shown below. 
![val_batch0_keypoint_pred](https://user-images.githubusercontent.com/47654539/169971653-8aab5ded-ff4c-4dc4-b0e0-286d1624c0ef.jpg)
But still failed with following OKSloss. Any ideas about this?
```
def calc_oksloss(p_kpt_x, p_kpt_y, gts, stride):
    sigmas = torch.tensor(np.array([.26, .25, .25, .35, .35, .79, .79, .72, .72, .62,.62, 1.07, 1.07, .87, .87, .89, .89])/10.0, device=gts.device)
    vars = (sigmas * 2)**2
    k = len(sigmas)
    ious = np.zeros(gts.shape[0])
    for i in range(gts.shape[0]): # each box keypoints
        # create bounds for ignore regions(double the gt bbox)
        g = gts[i]
        xg = g[4:38:2]; yg = g[5:38:2]; vg = g[38:55]; area = g[55]
        x0 = g[0] - g[2]*1.5; x1 = g[0] + g[2]*1.5   #extend bbox , cxcywh
        y0 = g[1] - g[3]*1.5; y1 = g[1] + g[3]*1.5
        k1 = torch.count_nonzero(vg > 0)     
        xd = p_kpt_x[i]; yd = p_kpt_y[i]
        if k1>0:
            # measure the per-keypoint distance if keypoints visible
            dx = xd - xg
            dy = yd - yg                
        else:
            # measure minimum distance to keypoints in (x0,y0) & (x1,y1)
            z = np.zeros((k))
            dx = np.max((z, x0-xd),axis=0)+np.max((z, xd-x1),axis=0)
            dy = np.max((z, y0-yd),axis=0)+np.max((z, yd-y1),axis=0)
        e = (dx**2 + dy**2) / vars / (area+np.spacing(1)) / 2
        if k1 > 0:
            e=e[vg > 0] 
        ious[i] = torch.sum(torch.exp(-e)) / e.shape[0]

    oksloss = 1 - ious.mean()
    return oksloss
```

**Comment by debapriyamaji** [comment]
@baoxin1100 I have released a preliminary version of the code. Can you please have a look at it.

https://github.com/TexasInstruments/edgeai-yolov5/tree/yolo-pose

Regards, Debapriya

**Comment by coordxyz** [comment]
@debapriyamaji  Thanks for releasing the code. Finally my code works with OKSloss. Here are my results.
![img](https://user-images.githubusercontent.com/47654539/173178756-39b0cf4c-ccf0-46d1-8f7b-94e6f5a97b9e.png)
![val_batch0_boxkeypoint_pred](https://user-images.githubusercontent.com/47654539/173178758-48180b25-052d-4317-b310-e76e502d96fd.jpg)
![val_batch0_keypoint_pred](https://user-images.githubusercontent.com/47654539/173178760-a612fd1c-0155-4905-b719-94efd1c35899.jpg)

**Comment by coordxyz** [comment]
@debapriyamaji I think the implementation of OKSloss is different with the description in the paper. Could you please explain why? Thx~                 
```
   #oks based loss
                    d = (pkpt_x-tkpt[i][:,0::2])**2 + (pkpt_y-tkpt[i][:,1::2])**2
                    s = torch.prod(tbox[i][:,:-2], dim=1, keepdim=True)
                    kpt_loss_factor = (torch.sum(kpt_mask != 0) + torch.sum(kpt_mask == 0))/torch.sum(kpt_mask != 0)
                    lkpt += kpt_loss_factor*((1 - torch.exp(-d/(s*(4*sigmas**2)+1e-9)))*kpt_mask).mean()

```
![image](https://user-images.githubusercontent.com/47654539/173178873-5b699cef-fc77-43c9-b128-c964b231904b.png)

**Comment by trungpham2606** [comment]
Hi @coordxyz Can you show me your compute_loss class ? or any changes you did for making it work? Seem this repo contains a lot of bugs

**Comment by debapriyamaji** [comment]
@coordxyz What difference do you exactly see between the OKS definition and the loss function in the code.

**Comment by a-dubbel** [comment]
Okay, so I think I understand now. 

 k = 2σ, so that's why k<sup>2</sup> = 4σ<sup>2</sup>. 
 
 But the code only has **s** instead of **2s**. Is that because your scale is the area of the bbox instead of segmentation area?

**Comment by Song-Jie-ing** [comment]
> Okay, so I think I understand now.
> 
> k = 2σ, so that's why k2 = 4σ2.
> 
> But the code only has **s** instead of **2s**. Is that because your scale is the area of the bbox instead of segmentation area?

Hello, may I ask the difference between kpt and kptv indicators
